### PR TITLE
Transcripts bug - transcripts from different sources of the same language do not switch

### PR DESCRIPTION
**Pull request by wowkalucky**

### CODE REVIEW

File: video_xblock/static/html/studio-edit-transcripts.html [hunk 1]
```diff
@@ -42,5 +42,5 @@
 </div>
 
 <form method="POST" action="/assets/{{ courseKey }}/" class="file-uploader-form is-hidden" enctype="multipart/form-data">
-  <input class="input-file-uploader" name="file" type="file" accept="" data-change-field-name="" data-lang-code="" data-lang-label="">
+  <input class="input-file-uploader" name="file" type="file" accept="" data-change-field-name="" data-lang-code="" data-lang-label="" {% for source in sources %}data-source-{{ source.0 }}="{{ source.1 }}"{% endfor %}>
```
Comment (z4y4ts) [inline video_xblock/static/html/studio-edit-transcripts.html]: Break this long line

File: video_xblock/backends/brightcove.py [hunk 2]
```diff
@@ -525,47 +529,46 @@ def get_default_transcripts(self, **kwargs):
...
+
+        # Handle empty response (no subs uploaded on a platform)
+        captions_data = text.get('text_tracks')
+
+        if not captions_data:
+            message = _("For now, video platform doesn't have any timed transcript for this video.")
+            return default_transcripts, message
+
+        # Populate default_transcripts
+        transcripts_data = [[cap_data.get('src'), cap_data.get('srclang')] for cap_data in captions_data]
```
Comment (z4y4ts) [inline video_xblock/backends/brightcove.py]: Make `transcripts_data` generator, not a list.

File: video_xblock/utils.py [hunk 3]
```diff
@@ -93,3 +96,33 @@ def remove_escaping(text):
         "&lt;": "<"
     }
     return unescape(text, html_unescape_table)
+
+
+def create_reference(lang_label, video_id, source="default"):
+    """
+    Build transcript file reference based on input information.
```
Comment (z4y4ts) [inline video_xblock/utils.py]: It looks like it should actually be `reference name` rather than just `reference`.

File: video_xblock/utils.py [hunk 4]
```diff
@@ -93,3 +96,33 @@ def remove_escaping(text):
...
+        source=source,
+    ).encode('utf8')
+    return reference
+
+
+def filter_transcripts_by_source(transcripts, source=TranscriptSource.DEFAULT):
+    """
+    Filter given transcripts by source attribute.
+    """
+    return [tr for tr in transcripts if tr['source'] == source]
```
Comment (z4y4ts) [inline video_xblock/utils.py]: Make it a generator, not a list.

File: video_xblock/video_xblock.py [hunk 5]
```diff
@@ -20,13 +20,15 @@
...
+from .constants import PlayerName, TranscriptSource
 from .exceptions import ApiClientError
 from .mixins import ContentStoreMixin, LocationMixin, PlaybackStateMixin, SettingsMixin, TranscriptsMixin
 from .workbench.mixin import WorkbenchMixin
 from .settings import ALL_LANGUAGES
 from .fields import RelativeTime
-from .utils import render_template, render_resource, resource_string, ugettext as _
+from .utils import (
+    render_template, render_resource, resource_string, ugettext as _, create_reference, filter_transcripts_by_source,
+    normalize_transcripts)
```
Comment (z4y4ts) [inline video_xblock/video_xblock.py]: Please sort `from .utils import (...)` alphabetically.

File: video_xblock/static/html/studio-edit-default-transcripts.html [hunk 6]
```diff
@@ -12,7 +12,7 @@
       <div class="custom-field-section-label enabled-transcripts">{% trans "Enabled transcripts" %}</div>
       {% if transcripts  %}
       <!-- Display a list of transcripts fetched from a video xblock -->
-        {% for transcript in transcripts %}
+        {% for transcript in enabled_default_transcripts %}
```
Comment (z4y4ts) [inline video_xblock/static/html/studio-edit-default-transcripts.html]: You're changing it here, but not on line 74.
How does this align?

File: video_xblock/static/js/studio-edit/transcripts-manual-upload.js [hunk 7]
```diff
@@ -130,7 +130,7 @@ function validateTranscriptFile(event, fieldName, filename, $fileUploader) {
  * @param {String} oldLang
  * @param {Array} transcriptsValue
  */
-function pushTranscript(lang, label, url, oldLang, transcriptsValue) {
+function pushTranscript(lang, label, url, source, oldLang, transcriptsValue) {
```
Comment (z4y4ts) [inline video_xblock/static/js/studio-edit/transcripts-manual-upload.js]: Please update JsDoc as well

File: video_xblock/video_xblock.py [hunk 8]
```diff
@@ -388,18 +391,21 @@ def studio_view(self, _context):
...
+            'advanced_fields': advanced_fields,
+            'auth_error_message': auth_error_message,
+            'basic_fields': basic_fields,
             'courseKey': self.course_key,
             'languages': languages,
+            'player_name': self.player_name,  # for players identification
+            'players': PlayerName,
+            'sources': TranscriptSource.to_dict().items(),
+            # transcripts context:
             'transcripts': transcripts,
```
Comment (z4y4ts) [inline video_xblock/video_xblock.py]: Do you still need `transcripts` in the `context`?

Comment (wowkalucky) [inline video_xblock/video_xblock.py]: We need `transcripts` in the context - it's a kind of "all enabled transcripts" variable.

File: video_xblock/static/html/studio-edit-default-transcripts.html (hunk 6 above)
Comment (wowkalucky) [inline video_xblock/static/html/studio-edit-default-transcripts.html]: Yes, my bad - here we don't need extra set.

File: video_xblock/tests/unit/test_video_xblock_handlers.py [hunk 9]
```diff
@@ -40,3 +41,58 @@ def test_auth_video_api_handler_delegates_call(self, auth_video_api_mock):
...
+
+class UploadDefaultTranscriptHandlerTests(VideoXBlockTestBase):
+    """
+    Test cases for `VideoXBlock.upload_default_transcript_handler`
+    """
+
+    @patch('video_xblock.video_xblock.create_reference_name')
+    def test_upload_handler_default_transcript_not_in_vtt_case(self, create_reference_name_mock):
+        # Arrange
+        request_body = """{"label": "test_label","lang": "test_lang","source": "test_source","url": "test_url"}"""
```
Comment (z4y4ts) [inline video_xblock/tests/unit/test_video_xblock_handlers.py]: You can use `'` instead of `"""` here.

**Pull request merged**